### PR TITLE
Update cleanup_html to format content correctly

### PR DESCRIPTION
**Pull request by rachkor**
Important formatting wasn't getting picked up by the old `cleanup_html`. This PR allows for links, code blocks, and lists to be parsed correctly. 

This PR has also been updated to resolve https://github.com/dradis/dradis-ce/issues/216 and adds the following new fields: 

- impact
- actionsToTake
- requiredSkillsForExploitation
- externalReferences
- remedyReferences
- extrainformation

### CODE REVIEW

File: lib/netsparker/vulnerability.rb [hunk 1]
```diff
@@ -105,19 +105,35 @@ def cleanup_html(source)
...
+      result.gsub!(/<p (.*?)>(.*?)<\/p>/) { "\n#{$2}\n" }
+      result.gsub!(/<span(.*?)>(.*?)<\/span>/, '\2')
+      result.gsub!(/(<p>)|(<\/p>)/, "\n")
+
+      result.gsub!(/<a href=\"(.*?)\" (.*?)>(.*?)<\/a>/i) { "\"#{$3.strip}\":#{$1.strip}" }
+
+      result.gsub!(/<code><pre.*?>(.*?)<\/pre><\/code>/) {"\n\nbc. #{$1}\n\n" }
+      result.gsub!(/<code><pre.*?>(.*?)<\/pre><\/code>/m) {|m| "\n\nbc.. #{$1}\n\np.  \n" }
+      
+      result.gsub!(/<pre.*?>(.*?)<\/pre>/) {"\n\nbc. #{$1}\n\n" }
```
Comment (dormi) [inline lib/netsparker/vulnerability.rb]: would it be possible to join the two `gsub('pre')` calls into a single one?

File: lib/netsparker/vulnerability.rb [hunk 2]
```diff
@@ -105,19 +105,35 @@ def cleanup_html(source)
...
+      result.gsub!(/<pre.*?>(.*?)<\/pre>/m) {|m| "\n\nbc.. #{$1}\n\np.  \n" }
+
+      result.gsub!(/<code>(.*?)<\/code>/) {"\n\nbc. #{$1}\n\n" }
```
Comment (dormi) [inline lib/netsparker/vulnerability.rb]: would it be possible to join the two `gsub('code')` calls into a single one?

File: lib/netsparker/vulnerability.rb [hunk 3]
```diff
@@ -105,19 +105,35 @@ def cleanup_html(source)
...
+      result.gsub!(/<code>(.*?)<\/code>/m) {|m| "\n\nbc.. #{$1}\n\np.  \n" }
+
+      result.gsub!(/(<ul>)|(<\/ul>|(<ol>)|(<\/ol>))/, "\n")
+      result.gsub!(/<li(.*?)>(.*?)<\/li>/) {"\n* #{$2}" }
```
Comment (dormi) [inline lib/netsparker/vulnerability.rb]: also, would it be possible to have a single `gsub('li')` ?

File: lib/netsparker/vulnerability.rb [hunk 4]
```diff
@@ -105,19 +105,35 @@ def cleanup_html(source)
...
-      result.gsub!(/<pre.*?>(.*?)<\/pre>/m){|m| "\n\nbc.. #{$1.strip}\n\np.  \n" }
-      result.gsub!(/<span.*>(.*?)<\/span>/, '\1')
-      result.gsub!(/<ul>(.*?)<\/ul>/m, '\1')
+      result.gsub!(/<p (.*?)>(.*?)<\/p>/) { "\n#{$2}\n" }
+      result.gsub!(/<span(.*?)>(.*?)<\/span>/, '\2')
+      result.gsub!(/(<p>)|(<\/p>)/, "\n")
+
+      result.gsub!(/<a href=\"(.*?)\" (.*?)>(.*?)<\/a>/i) { "\"#{$3.strip}\":#{$1.strip}" }
+
+      result.gsub!(/<code><pre.*?>(.*?)<\/pre><\/code>/) {"\n\nbc. #{$1}\n\n" }
```
Comment (dormi) [inline lib/netsparker/vulnerability.rb]: may be we can merge the `gsub('<code><pre>')` calls?
Or maybe merge all `gsub` calls that transform xml with `code` and `pre` could be handled by a single `gsub` call?

File: lib/netsparker/vulnerability.rb [hunk 5]
```diff
@@ -105,19 +105,35 @@ def cleanup_html(source)
...
+      result.gsub!(/<div>(.*?)<\/div>/m, '\1')
       result.gsub!(/<font.*?>(.*?)<\/font>/m, '\1')
-      result.gsub!(/<h2>(.*?)<\/h2>/) { "*#{$1.strip}*" }
-      result.gsub!(/<i>(.*?)<\/i>/, '\1')
-      result.gsub!(/<li.*>(.*)?<\/li>/, '* \1')
-      result.gsub!(/<p>(.*?)<\/p>/, '\1 ')
-      result.gsub!(/<pre.*?>(.*?)<\/pre>/m){|m| "\n\nbc.. #{$1.strip}\n\np.  \n" }
-      result.gsub!(/<span.*>(.*?)<\/span>/, '\1')
-      result.gsub!(/<ul>(.*?)<\/ul>/m, '\1')
+      result.gsub!(/<p (.*?)>(.*?)<\/p>/) { "\n#{$2}\n" }
```
Comment (dormi) [inline lib/netsparker/vulnerability.rb]: merge `gsub('p')` calls in a single call?

**Comment by rachkor** [comment]
@dormi sorry for the delay here. The `gsub`s are complex and plentiful because Netsparker likes to play fast and loose with their formatting. I tried to merge as much as I could but this was the combo that got everything working as expected. I'd like to merge this now and we can refactor it later if needed.

File: templates/issue.template [hunk 6]
```diff
@@ -11,5 +11,4 @@
 %issue.description%
 
 #[Remedy]#
-%issue.remedy%
-
+%issue.remedy%
```
Comment (etdsoft) [inline templates/issue.template]: can we address this before merging?

File: templates/issue.fields [hunk 7]
```diff
@@ -15,7 +16,12 @@ issue.classification_pci31
...
+issue.external_references
+issue.extrainformation
+issue.impact
 issue.remedy
+issue.remedy_references
+issue.required_skills_for_exploitation
 issue.severity
 issue.title
-issue.type
+issue.type
```
Comment (etdsoft) [inline templates/issue.fields]: we've lost and end-of-file new line here.

File: lib/netsparker/vulnerability.rb [hunk 8]
```diff
@@ -23,7 +23,7 @@ def supported_tags
         :title,
 
         # simple tags
-        :certainty, :description, :rawrequest, :rawresponse, :remedy, :severity,
+        :actions_to_take, :certainty, :description, :external_references, :extrainformation, :impact, :rawrequest, :rawresponse, :remedy, :remedy_references, :required_skills_for_exploitation, :severity,
```
Comment (etdsoft) [inline lib/netsparker/vulnerability.rb]:
```suggestion
        :actions_to_take, :certainty, :description, :external_references,
        :extrainformation, :impact, :rawrequest, :rawresponse, :remedy,
        :remedy_references, :required_skills_for_exploitation, :severity,
```

File: lib/netsparker/vulnerability.rb [hunk 9]
```diff
@@ -118,27 +122,42 @@ def cleanup_html(source)
...
+      result.gsub!(/<li>/, "\n* ")
+      result.gsub!(/<\/li>/, "\n")
 
       result
     end
 
     # Some of the values have embedded HTML conent that we need to strip
     def tags_with_html_content
-      [:description, :remedy]
+      [:actions_to_take, :description, :external_references, :extrainformation, :impact, :remedy, :remedy_references, :required_skills_for_exploitation]
```
Comment (etdsoft) [inline lib/netsparker/vulnerability.rb]:
```suggestion
      [
        :actions_to_take, :description, :external_references, :extrainformation,
        :impact, :remedy, :remedy_references, :required_skills_for_exploitation
      ]
```

File: templates/issue.sample [hunk 10]
```diff
@@ -4,7 +4,13 @@
   <severity>Information</severity>
   <certainty>100</certainty>
   ​<description><![CDATA[<p>Netsparker detected a missing <code>X-XSS-Protection</code> header which means that this website could be at risk of a Cross-site Scripting (XSS) attacks.</p>]]></description>
+  <impact><![CDATA[ <div>;This issue is reported as additional information only. There is no direct impact arising from this issue.</div> ]]></impact>
```
Comment (etdsoft) [inline templates/issue.sample]: the sample is missing `<extrainformation>`, no?

File: lib/dradis/plugins/netsparker/gem_version.rb [hunk 11]
```diff
@@ -9,7 +9,7 @@ def self.gem_version
       module VERSION
         MAJOR = 3
         MINOR = 10
```
Comment (etdsoft) [inline lib/dradis/plugins/netsparker/gem_version.rb]:
```suggestion
        MINOR = 11
```

File: lib/dradis/plugins/netsparker/gem_version.rb [hunk 12]
```diff
@@ -9,7 +9,7 @@ def self.gem_version
...
-        TINY = 0
+        TINY = 1
```
Comment (etdsoft) [inline lib/dradis/plugins/netsparker/gem_version.rb]:
```suggestion
        TINY = 0
```

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,6 +1,9 @@
 ## Dradis Framework 3.10 (August, 2018) ##
 
 *   Add CVSS fields for Issues.
+*   Add impact, actionsToTake, requiredSkillsForExploitation, externalReferences, remedyReferences, and extrainformation fields
```
Comment (etdsoft) [inline CHANGELOG.md]: I don't think this made it into 3.10?

Also 3.10 wasn't released in Aug unfortunately.

We probably need to update the 3.10 release date and notes, and add a 3.11 section.

Comment (rachkor) [reply CHANGELOG.md]: Updated!

File: lib/dradis/plugins/netsparker/gem_version.rb (hunk 11 above)
Comment (rachkor) [reply lib/dradis/plugins/netsparker/gem_version.rb]: Updated!

File: templates/issue.sample (hunk 10 above)
Comment (rachkor) [reply templates/issue.sample]: Thank you!

**Comment by rachkor** [comment]
Fixed everything (I think!)

**Comment by dormi** [comment]
I merged this into master.
Then reverted it, and recreated the PR: https://github.com/dradis/dradis-netsparker/pull/9

**Pull request merged**